stdlib/asm/collections/smt.masm

```diff
@@ -0,0 +1,163 @@
+#! input:  [v, ...]
+#! output: [v >> 48, ...]
+proc.shr_48
+    u32split
+    push.65536
+    mul
+    u32split
+    swap
+    drop
+    swap
+    drop
+end
+
+#! input:  [v, ...]
+#! output: [v >> 32, ...]
+proc.shr_32
+    u32split
+    swap
+    drop
+end
+
+#! input:  [v, ...]
+#! output: [v >> 16, ...]
+proc.shr_16
+    u32split
+    push.65536
+    mul
+    u32split
+    push.4294967296
+    mul
+    swap.2
+    push.65536
+    mul
+    u32split
+    swap
+    drop
+    add
+    add
+end
+
+#! Fetches the keyed value of a Sparse Merkle tree with the provided root.
+#!
+#! `V` will be `0` if there is no leaf value for `K` in `R`.
+#!
+#! input:  [K, R, ...]
+#! output: [V, R, ...]
+export.get
+    # invoke adv and fetch target depth
+    # [d, K, R, ...]
+    adv.smtget adv_push.1
+
+    # if has value, compute node value; use empty root otherwise
+    # [N, d, K, R, V, ...]
+    adv_push.1
+    if.true
+        # make a domain word for depth
+        # [0, 0, d, 0, K, R, ...]
+        push.0 swap push.0.0
+
+        # load the value
+        # [V, 0, 0, d, 0, K, R, V, ...]
+        padw adv_loadw movdnw.3 dupw.3
+
+        # compute remaining path key
+        # [K', V, 0, 0, d, 0, K, R, V, ...]
+        dupw.2 dup.10 eq.16
+        if.true
+            exec.shr_16
+        else
+            dup.10 eq.32
+            if.true
+                exec.shr_32
+            else
+                dup.10 eq.48
+                if.true
+                    exec.shr_48
+                else
+                    dup.10 eq.64
+                    if.true
+                        # remaining path of depth 64 is 0
+                        drop push.0
+                    else
+                        # adv provider returned an invalid depth
+                        # panic
+                        push.0
+                        assert
+                    end
+                end
+            end
+        end
```

Comment:
We should be able to simplify this to do it without conditional statements. Basically, what we need to do here is zero out $d$ most significant bits of the key where $d$ is the depth of the node. I think we should be able to do this by relying on `exp` and `u32unchecked_divmod` instructions.

Reply:
This might be more difficult than I originally thought (because we also need to handle the case for depth being 64). A simpler alternative while still using conditional statements may be:
```
adv_push.2 # put 2 flags onto the stack to control branching
if.true
  if.true
    # assert depth is 16 and process this case
  else
    # assert depth is 32 and process this case
  end
else
  if.true
    # assert depth is 48 and process this case
  else
    # assert depth is 64 and process this case
  end
end
```

Reply:
Fixed! I also removed the conditional from the concrete implementations by increasing a bit the instructions